chore: transition the library to the new microgenerator

plamut opened the pull request:
Closes #131.
Closes #168.

This PR replaces the old code generator, the generated parts of the code now use microgenerator. The latter also implies dropping support for Python 2.7 and 3.5.

There are a lot of changes and the transition itself was far from smooth, thus it's probably best to review this commit by commit (I tried to make the commits self-contained with each containing a single change/fix).

### Things to focus on in reviews
 - Regenerating the code overrides some of the URLs in the samples README. Seems like a synthtool [issue](https://github.com/googleapis/synthtool/issues/663).
 - The clients do not support the `client_config` argument anymore. At least the ordering keys feature used that to change the timeout to "infinity". We need to see if that's crucial, and if the same can be set in a different way (maybe through the retry policy...)
 - `SERVICE_ADDRESS` and `_DEFAULT_SCOPES` constants in clients might be obsolete. Let's see if there are more modern alternatives (they are currently still injected into the generated clients).
- Regenerating the code out of the box fails, because the `Bazel` tool incorrectly tries to use Python 2, resulting in syntax errors (could be just a problem on my machine, but it is a known Bazel [issue](https://github.com/bazelbuild/bazel/issues/4815)).

  **Workaround:**
    - Add the snippet from the [comment](https://github.com/bazelbuild/bazel/issues/4815#issuecomment-460777113) to `google/pubsub/v1/BUILD.bazel` file in the local [googleapis](https://github.com/googleapis/googleapis) clone and point the `synthtool` to it:
    ```
    $ export SYNTHTOOL_GOOGLEAPIS=/path/to/googleapis
    ```
    - Patch the local `synthtool` installation.  Add the following two Bazel arguments to the `_generate_code()` method in `synthtool/gcp/gapic_bazel.py` (lines 177-178):
        ```py
        "--python_top=//google/pubsub/v1:myruntime",
        "--incompatible_use_python_toolchains=false",
        ```
  The workaround should convince Bazel to use Python 3, as this is the Python version in the configs.

### Things left to do
 - [x] ~~**Double check that the new client performance is adequate. There have been reports of possibly degraded performance with the new microgenerator.**~~ The 10% performance hit was declared acceptable, not a release blocker anymore.
 - [x] After approvals, re-generate the code one more time to make sure it works without errors (such as a new version of black wanting to re-generate some files and causing the CI check to fail).
 - [x] Lint samples. It appears that the "fixup_keywords" step in the migration guide made the linter unhappy. :smile: 
 - [x] Fix samples and their tests.
 - [x] Fix system tests.
 - [x] Determine how to handle methods that are now either missing, e.g.  `get_iam_policy()`, or do not support all config options anymore, e.g `create_subscription()`. Adjust or delete?
 - [x] ~~Determine a replacement for now-unsupported `client_config` argument to client constructor. Or does the code generator need an update?~~   `client_config` has been replaced by custom Retry settings passed to the GAPIC client `publish()`. If we want to support custom retries, we must update the user-facing client's `publish()` method.
 - [x] Add UPGRADING guide to the docs (also depends on the previous point - some of the changes might actually be bugs in the new generator).
 - [ ] Make the samples CI checks required?

### PR checklist
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-pubsub/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

plamut commented:
I believe this PR is now ready to be reviewed. The things that still fail are related to incompatible changes in the client itself, for example:
 - Missing IAM methods - `get_iam_policy()`, `set iam policy()`, `test_iam_permission()`
 - Missing `create_subscription()` parameters - `dead_letter_policy`, `expiration_policy`, etc.
 - Missing `client_config` parameter in publisher client constructor (is there an alternative mechanism?)

We need to figure out how to handle these and I will add the `UPGRADING.md` file to the docs after all these things have been clarified.

Since this PR is quite complex, every pair of eyes would be beneficial. :slightly_smiling_face: 

 - @pradn Please check if ordering keys are affected in any way due to the lack of `client_config` options - they need "infinite" timeouts, right?
 - @anguillanneuf Any feedback on samples would be great, as you're probably most familiar with them.
 - @busunkim96 General feedback on the migration process in case I missed something.

Let's not rush and take the time to review this thoroughly, thanks!

anguillanneuf commented:
@plamut Samples update should be added in a separate PR because samples are tested against the published library of the version defined in `samples/snippets/requirements.txt`. We can land this PR first, cut a release, then update samples.

busunkim96 commented:
@anguillanneuf The presubmit and continuous builds (for samples) install the library from source. There's an additional periodic build that uses the latest published version of the library. I think the failing sample tests are from the issues @plamut identified above.

That said once this is merged we should probably follow quickly with the release PR, to avoid confusing folks.

plamut commented:
@anguillanneuf We can set the `INSTALL_LIBRARY_FROM_SOURCE` environment variable and the samples will be installed from source, testing against the very latest version there exists. This is the key check in samples noxfile:

https://github.com/googleapis/python-pubsub/pull/158/files#diff-1431c4677d4eb44d5cf331ecf9b3657cR160-R161

anguillanneuf commented on samples/snippets/subscriber_test.py:
```diff
@@ -235,17 +253,17 @@ def test_delete(subscriber_client, subscription_admin):
 
 
-def _publish_messages(publisher_client, topic):
+def _publish_messages(publisher_client, topic, **attrs):
```
👍

anguillanneuf commented:
@plamut @busunkim96 Ah I see. Immediate new release sounds good to me. Thank you for confirming! The sample updates look right 👍 . I see pros and cons in using requests as arguments, but it's already a consensus reached by the microgenerator team, right? 
Will review again when IAM-related methods https://github.com/googleapis/gapic-generator-python/issues/536 and other issues that Peter has summarized in https://github.com/googleapis/gapic-generator-python/issues/530 are resolved. Thank you!

busunkim96 commented:
@anguillanneuf Yep, it ended up being the best way for us to make sure non breaking API changes at the proto level don't produce breaking changes at the library level.

plamut commented:
Since the code generator now adds IAM-related methods, I adjusted related system and samples tests, and these tests should now pass.

Docstring format in IAM-related methods will need some tweaking in `synth.py` to avoid warnings when generating the docs.

plamut commented on samples/snippets/README.rst.in:
```diff
@@ -29,4 +29,25 @@ samples:
\ No newline at end of file
+folder: samples/snippets
+
+metadata:
+  repo:
+    name_pretty: Google Cloud Pub/Sub
+    product_documentation: TODO (docs URL)
+    samples:
+    - name: Quickstart (Publisher)
+      description: TODO?
+      runnable: true
```
What is supposed to be here? What samples should be listed?

I added these dummy placeholders to avoid errors on synthtool runs, but we still need to replace them with correct values.

anguillanneuf replied:
description: This is the quickstart using in https://cloud.google.com/pubsub/docs/quickstart-py-mac.

busunkim96 replied:
It looks like this was a bug in the synthtool templates. Some folks are working on adding a new README format, but we're not quite ready to use it yet. Kristen merged a PR to exclude the new README from the template group, so you should be able to remove this "metadata" section. https://github.com/googleapis/synthtool/pull/685

plamut replied:
Sounds good, thanks!

plamut replied:
Removed that commit, the fix is not needed anymore with the latest synthtool.

plamut commented:
@pradn Setting the total retry timeout to "infinite" seems to be the only thing left failing. It's done by customizing the `client_config` argument to publisher client, but that argument is not supported anymore by the new generated client.
https://github.com/googleapis/python-pubsub/blob/b1bfc9e7ae540fed4a5ae81f9d895072a255a15d/google/cloud/pubsub_v1/publisher/client.py#L173-L191

Since retry settings/timeouts need to be configured differently now, do you remember which are the key methods that needs to be configured for the ordering keys to work? Is it only the `publish()` method that needs to have its timeout/retry settings changed?

plamut commented on samples/snippets/publisher.py:
```diff
@@ -212,65 +212,6 @@ def callback(future):
     # [END pubsub_publisher_batch_settings]
 
 
-def publish_messages_with_retry_settings(project_id, topic_id):
-    """Publishes messages with custom retry settings."""
-    # [START pubsub_publisher_retry_settings]
```
If this is still needed, we must add support for custom retries to the `client.publish()` method and override the default publish retry settings. Ordering keys will still override the deadline to "infinity", though, which should be documented.

plamut replied:
Not relevant anymore, added support for custom retries.

plamut commented:
@pradn @anguillanneuf @busunkim96 The relevant CI checks are all green now, this can now be reviewed. Any pair of eyes is welcome if you have capacity.

plamut commented:
@busunkim96 Side question - I added the UPGRADING guide to the 2.x versions, but the generated code is placed in the `google/pubsub_v1/services/` directory - shouldn't it be generated in `pubsub_v2` instead?

I can be confusing if the types live in the `pubsub_v1` path according to the docs, e.g. `google.cloud.pubsub_v1.types.AcknowledgeRequest`, while the library version is actually `2.x`.

busunkim96 commented:
@plamut The library version doesn't always match the major version of the API. Maybe we can tweak the wording in the migration guide to make that more clear?

plamut commented:
@busunkim96 I see, let's tweak the guide then. Thanks!

anguillanneuf commented on samples/snippets/publisher.py:
```diff
@@ -217,56 +215,38 @@ def callback(future):
-    publisher = pubsub_v1.PublisherClient(client_config=retry_settings)
+    # Configure the retry settings.
+    custom_retry = api_core.retry.Retry(
+        initial=0.250,  # seconds
+        maximum=90.0,  # seconds
+        multiplier=1.45,
+        deadline=300.0,  # seconds
```
Add default settings in comments.

anguillanneuf commented on scripts/readme-gen/templates/install_deps.tmpl.rst:
```diff
@@ -5,14 +5,14 @@ Install Dependencies
 
-        $ git clone https://github.com/googleapis/python-pubsub.git
+        $ git clone https://github.com/GoogleCloudPlatform/python-docs-samples.git
 
 #. Install `pip`_ and `virtualenv`_ if you do not already have them. You may want to refer to the `Python Development Environment Setup Guide`_ for Google Cloud Platform for instructions.
 
    .. _Python Development Environment Setup Guide:
        https://cloud.google.com/python/setup
 
-#. Create a virtualenv. Samples are compatible with Python 3.6+.
+#. Create a virtualenv. Samples are compatible with Python 2.7 and 3.4+.
```
This isn't the real output, right?

plamut replied:
This is a template ... but it seems that synth still overrides it with old values, I'll see if these changes should be reverted. It's more important what ends up in the actual README, etc. docs.

IIRC there is already a synthtool ticket to prevent this from happening in the first place.

plamut replied:
Fixed that (again), and there is a [ticket](https://github.com/googleapis/synthtool/issues/663) to fix this permanently in synthtool.

plamut commented:
**Update:**
Re-generated the code to include the latest changes, bumped dependencies to their latest versions and ran the benchmark again. It appears that even with the update the throughput performance has not been improved, unfortunately.
```
INFO-===============================================
1202724 [main] INFO com.google.pubsub.flic.Driver  - ===============================================
INFO-Results for cps-gcloud-python-publisher:
1202724 [main] INFO com.google.pubsub.flic.Driver  - Results for cps-gcloud-python-publisher:
INFO-50%: 5.0625 - 7.59375
1202728 [main] INFO com.google.pubsub.flic.Driver  - 50%: 5.0625 - 7.59375
INFO-99%: 86.49755859375 - 129.746337890625
1202728 [main] INFO com.google.pubsub.flic.Driver  - 99%: 86.49755859375 - 129.746337890625
INFO-99.9%: 194.6195068359375 - 291.92926025390625
1202728 [main] INFO com.google.pubsub.flic.Driver  - 99.9%: 194.6195068359375 - 291.92926025390625
INFO-Average throughput: 53.82 MB/s
1202728 [main] INFO com.google.pubsub.flic.Driver  - Average throughput: 53.82 MB/s
INFO-Results for cps-gcloud-python-subscriber:
1202728 [main] INFO com.google.pubsub.flic.Driver  - Results for cps-gcloud-python-subscriber:
INFO-50%: 127834.03948858939 - 191751.0592328841
1202729 [main] INFO com.google.pubsub.flic.Driver  - 50%: 127834.03948858939 - 191751.0592328841
INFO-99%: 191751.0592328841 - 287626.58884932613
1202729 [main] INFO com.google.pubsub.flic.Driver  - 99%: 191751.0592328841 - 287626.58884932613
INFO-99.9%: 287626.58884932613 - 431439.8832739892
1202729 [main] INFO com.google.pubsub.flic.Driver  - 99.9%: 287626.58884932613 - 431439.8832739892
INFO-Average throughput: 38.17 MB/s
1202729 [main] INFO com.google.pubsub.flic.Driver  - Average throughput: 38.17 MB/s
INFO-===============================================
```

busunkim96 commented:
CC @software-dov 

@plamut Do you have an idea as to where the slowdown might be coming from? Which path(s) are being tested by the load test framework?

plamut commented:
@busunkim96 I didn't have the time (yet) to profile the client in-depth. It seems, though, that the load test framework more or less uses the `publish()` and `streaming_pull()` methods. The publisher/subscriber code that is run can be found [here](https://github.com/GoogleCloudPlatform/pubsub/tree/master/load-test-framework/python_src/clients) (in `cps_publisher_task.py` and  `cps_subscriber_task.py`, both are pretty short).

The only thing I changed between the test runs was a [line](https://github.com/GoogleCloudPlatform/pubsub/blob/master/load-test-framework/python_src/requirements.txt#L2) in `requirements.txt`, changed the pubsub dependency to point to the PR branch:
```
git+https://github.com/plamut/python-pubsub@use-microgenerator#egg=google-cloud-pubsub
```

software-dov commented:
I've got no immediate hypotheses, and my only suggestion at this point is to gather cProfile data for analysis.

software-dov commented:
I don't think I understand the benchmark result very well; are the two different throughput numbers for `cps_publisher_task.py` and `cps_subscriber_task.py` respectively?

Also, just had a thought that may be relevant: is the manual layer invoking the asynchronous client or the synchronous? I see a lot of task/future names and semantics, which made me ask. The asynchronous surface has had basically no optimization focus.

plamut commented:
I did some publisher profiling with `yappi` using roughly [the code](https://github.com/GoogleCloudPlatform/pubsub/blob/master/load-test-framework/python_src/clients/cps_publisher_task.py#L55-L68) the benchmark framework uses and it seems that in the main thread a lot of time (~70%) is spent constructing `PubsubMessage` instances inside the `publisher.publish()` method:
```py
from google.pubsub_v1 import types as gapic_types

...
# Create the Pub/Sub message object.
message = gapic_types.PubsubMessage(
    data=data, ordering_key=ordering_key, attributes=attrs
)
```

FWIW, profiling the current released version (i.e. the `master`) branch did now show that piece of code to be problematic.

Is constructing a `PubsubMessage` more expensive with the microgenerator and are there ways to construct it faster, e.g. by circumventing any checks/extra logic that might be redundant for this use case?

software-dov commented:
Yes, constructing messages is more expensive with the microgenerator because of its use of [proto-plus](https://proto-plus-python.readthedocs.io/). Depending on what `data`, `ordering_key`, and `attrs` (actual params, not formal) are, proto-plus may be doing non-trivial work to marshal python structures to protobuf structures.

Can you send me the raw data in some way? I'd like to take a closer look at which part of construction is expensive.
There's a potential hack to get around this, but it's not really something we can recommend our users do.
It's possible to interact with 'vanilla' proto types in Python and then just shove them into proto-plus types before passing the result into a method call.

E.g.
```python
vanilla_pb = gapic_types.PubsubMessage.pb(data=data, ordering_key=ordering_key, attributes=attrs)
proto_plus_pb = gapic_types.PubsubMessage.wrap(vanilla_pb)
```

software-dov commented:
Optimizing `marshal.to_proto` is a mid-low priority ongoing effort. The good news is that any changes made to proto-plus will be available to all users for free and transparently, so if the new throughput numbers are acceptable they may eventually get better without having to cut a new release.
The not so good news is that I don't think `to_proto` is likely to get much faster in the general case. There may be specific instances we can optimize, or some crazy fast-path shenanigans, but the code there is doing nonzero work in order to facilitate a more flexible and ergonomic interface. The downside of proto-plus is that its user visible benefits come with a performance cost.

software-dov commented:
My bad, when I said "the raw data" above, I meant the profiling data. I don't particularly want to set up the benchmark environment, but I do want to try looking through the data to see what the hot spots are and if they could be optimized.

software-dov commented:
I have a special, optimized-but-not-reviewed-or-production-ready tarball of proto-plus. Is it possible to run the benchmark suite using it? My own benchmarking has these changes saving about 10-15 milliseconds out of about 500 milliseconds. YMMV, but if it improves the throughput enough it could be worth putting the proto-plus changes up for review.
[proto-plus-1.7.1.tar.gz](https://github.com/googleapis/python-pubsub/files/5120188/proto-plus-1.7.1.tar.gz)

plamut commented:
> I have a special, optimized-but-not-reviewed-or-production-ready tarball of proto-plus. Is it possible to run the benchmark suite using it?

If `pip` can access and install it, we can specify this version as a dependency in the framework's [requirements.txt](https://github.com/GoogleCloudPlatform/pubsub/blob/master/load-test-framework/python_src/requirements.txt), should be doable.

Just to clarify, would you like to run the benchmark using the tip of this PR branch, i.e. the version that already includes the optimizations that try to bypass the protobuf buffer classes? Or on the version that excludes this last commit and uses the wrapper classes more heavily?

> My bad, when I said "the raw data" above, I meant the profiling data.

No worries, sent you an email with the data just now.

software-dov commented:
The tip of this PR branch.

Changes are also available in this branch of my fork: https://github.com/software-dov/proto-plus-python/tree/hackaday

plamut commented:
First run, using proto-plus 1.7.1. The results seem more or less similar to the previous benchmark:
```
INFO-Results for cps-gcloud-python-publisher:
1151295 [main] INFO com.google.pubsub.flic.Driver  - Results for cps-gcloud-python-publisher:
INFO-50%: 5.0625 - 7.59375
1151304 [main] INFO com.google.pubsub.flic.Driver  - 50%: 5.0625 - 7.59375
INFO-99%: 86.49755859375 - 129.746337890625
1151304 [main] INFO com.google.pubsub.flic.Driver  - 99%: 86.49755859375 - 129.746337890625
INFO-99.9%: 194.6195068359375 - 291.92926025390625
1151304 [main] INFO com.google.pubsub.flic.Driver  - 99.9%: 194.6195068359375 - 291.92926025390625
INFO-Average throughput: 56.58 MB/s
1151306 [main] INFO com.google.pubsub.flic.Driver  - Average throughput: 56.58 MB/s
INFO-Results for cps-gcloud-python-subscriber:
1151306 [main] INFO com.google.pubsub.flic.Driver  - Results for cps-gcloud-python-subscriber:
INFO-50%: 1477.8918800354004 - 2216.8378200531006
1151306 [main] INFO com.google.pubsub.flic.Driver  - 50%: 1477.8918800354004 - 2216.8378200531006
INFO-99%: 25251.16829404235 - 37876.75244106352
1151307 [main] INFO com.google.pubsub.flic.Driver  - 99%: 25251.16829404235 - 37876.75244106352
INFO-99.9%: 37876.75244106352 - 56815.128661595285
1151307 [main] INFO com.google.pubsub.flic.Driver  - 99.9%: 37876.75244106352 - 56815.128661595285
INFO-Average throughput: 57.37 MB/s
1151307 [main] INFO com.google.pubsub.flic.Driver  - Average throughput: 57.37 MB/s
```

I also did another run, the measured throughput was around 55 MB/s - a bit worse, but within the normal results variance (the typical difference appears to be somewhere in the 1-2 MB/s range).

software-dov commented:
Okay. In light of those numbers I'm inclined not to merge the proto-plus changes. Is there an approval process for accepting the throughput regression?

plamut commented:
Please just mind that that the proto-plus optimizations might have noticeable effect here, as the optimizations added here actively try to circumvent proto-plus, but other libraries might still see benefits.

I'm not aware of any formal processes, but we do have weekly PubSub meetings on Thursdays where we discuss these things. I added the question to the agenda whether a -10% performance hit is acceptable in the new major release.

plamut commented:
@cguardia That's fine, appreciated. I will re-generate and benchmark the code again now, just in case, and see if we are ready for the major release.

plamut commented:
After re-generating the code yet again, performance still appears to be in line with the previous benchmarks:

```
INFO-Results for cps-gcloud-python-publisher:
1095535 [main] INFO com.google.pubsub.flic.Driver  - Results for cps-gcloud-python-publisher:
INFO-50%: 5.0625 - 7.59375
1095545 [main] INFO com.google.pubsub.flic.Driver  - 50%: 5.0625 - 7.59375
INFO-99%: 86.49755859375 - 129.746337890625
1095546 [main] INFO com.google.pubsub.flic.Driver  - 99%: 86.49755859375 - 129.746337890625
INFO-99.9%: 194.6195068359375 - 291.92926025390625
1095546 [main] INFO com.google.pubsub.flic.Driver  - 99.9%: 194.6195068359375 - 291.92926025390625
INFO-Average throughput: 56.04 MB/s
1095547 [main] INFO com.google.pubsub.flic.Driver  - Average throughput: 56.04 MB/s
INFO-Results for cps-gcloud-python-subscriber:
1095547 [main] INFO com.google.pubsub.flic.Driver  - Results for cps-gcloud-python-subscriber:
INFO-50%: 194.6195068359375 - 291.92926025390625
1095549 [main] INFO com.google.pubsub.flic.Driver  - 50%: 194.6195068359375 - 291.92926025390625
INFO-99%: 16834.112196028233 - 25251.16829404235
1095549 [main] INFO com.google.pubsub.flic.Driver  - 99%: 16834.112196028233 - 25251.16829404235
INFO-99.9%: 37876.75244106352 - 56815.128661595285
1095550 [main] INFO com.google.pubsub.flic.Driver  - 99.9%: 37876.75244106352 - 56815.128661595285
INFO-Average throughput: 56.33 MB/s
1095550 [main] INFO com.google.pubsub.flic.Driver  - Average throughput: 56.33 MB/s
```

Merging. :tada:

kamalaboulhosn reviewed (approved):
The upgrade guide looks good.
